File: src/main/java/cpsc224/items/weapons/WeaponFactory.java
package cpsc224.items.weapons;

import cpsc224.items.weapons.moves.Move;
import cpsc224.items.weapons.moves.MoveFactory;
import cpsc224.utils.BufferedImageBuilder;

import javax.swing.*;

/**
 * A factory class to create weapons.
 */
public class WeaponFactory {

    private static final int ICON_WIDTH = 32;
    private static final int ICON_HEIGHT = 32;

    private static ImageIcon getIcon(int row, int col) {
        BufferedImageBuilder imageBuilder = new BufferedImageBuilder("/sprites/items/items.png");
        return imageBuilder
                .sliceToSprite(ICON_WIDTH, ICON_HEIGHT, row, col)
                .scale(64, 64)
                .toImageIcon();
    }

    private static ImageIcon getIcon(String name) {
        BufferedImageBuilder imageBuilder = new BufferedImageBuilder("/images/Weapons/" + name + ".png");
        return imageBuilder
                .scale(64, 64)
                .toImageIcon();
    }

    // *** CREATURE-SPECIFIC WEAPONS ***

    public static Weapon createRatClaws() {
        final String name = "Rat Claws";
        final int tier = 1;
        final Move move = MoveFactory.createScratchMove();

        return new Weapon(name, tier, move, getIcon(2, 4));
    }

    public static Weapon createMushroomCap() {
        final String name = "Mushroom Cap";
        final int tier = 1;
        final Move move = MoveFactory.createSporeShotMove();

        return new Weapon(name, tier, move, getIcon(name));
    }

    public static Weapon createBirdTalons() {
        final String name = "Bird Talons";
        final int tier = 1;
        final Move move = MoveFactory.createScratchMove();

        return new Weapon(name, tier, move);
    }

    // *** OTHER WEAPONS ***

    public static Weapon createCrudeBow() {
        final String name = "Crude Bow";
        final int tier = 1;
        final Move move = MoveFactory.createShootMove();

        return new Weapon(name, tier, move, getIcon(9, 2));
    }

    public static Weapon createLongBow() {
        final String name = "Long Bow";
        final int tier = 2;
        final Move move = MoveFactory.createShootMove();

        return new Weapon(name, tier, move, getIcon(9, 3));
    }

    public static Weapon createCrossbow() {
        final String name = "Crossbow";
        final int tier = 3;
        final Move move = MoveFactory.createShootMove();

        return new Weapon(name, tier, move, getIcon(9, 4));
    }

    public static Weapon createToxicStaff() {
        final String name = "Toxic Staff";
        final int tier = 3;
        final Move move = MoveFactory.createToxicBoltMove();

        return new Weapon(name, tier, move, getIcon(10, 2));
    }

    public static Weapon createHealStaff() {
        final String name = "Heal Staff";
        final int tier = 3;
        final Move move = MoveFactory.createHealMove();

        return new Weapon(name, tier, move, getIcon(10, 1));
    }

    public static Weapon createRustyDagger() {
        final String name = "Rusty Dagger";
        final int tier = 1;
        final Move move = MoveFactory.createSlashMove();

        return new Weapon(name, tier, move, getIcon(0, 0));
    }

    public static Weapon createSteelSword() {
        final String name = "Steel Sword";
        final int tier = 2;
        final Move move = MoveFactory.createSlashMove();

        return new Weapon(name, tier, move, getIcon(0, 2));
    }

    public static Weapon createRoyalSword() {
        final String name = "Royal Sword";
        final int tier = 3;
        final Move move = MoveFactory.createSlashMove();

        return new Weapon(name, tier, move, getIcon(0, 8));
    }

    public static Weapon createWoodClub() {
        final String name = "Wood Club";
        final int tier = 1;
        final Move move = MoveFactory.createSmashMove();

        return new Weapon(name, tier, move, getIcon(8, 2));
    }

    public static Weapon createSteelMace() {
        final String name = "Steel Mace";
        final int tier = 2;
        final Move move = MoveFactory.createSmashMove();

        return new Weapon(name, tier, move, getIcon(5, 0));
    }

    public static Weapon createSteelHammer() {
        final String name = "Steel Hammer";
        final int tier = 3;
        final Move move = MoveFactory.createSmashMove();

        return new Weapon(name, tier, move, getIcon(4, 4));
    }

    public static Weapon createRudimentaryStaff() {
        final String name = "Rudimentary Staff";
        final int tier = 1;
        final Move move = MoveFactory.createMagicMissileMove();

        return new Weapon(name, tier, move, getIcon(10, 0));
    }

    public static Weapon createEnchantedStaff() {
        final String name = "Enchanted Staff";
        final int tier = 2;
        final Move move = MoveFactory.createMagicMissileMove();

        return new Weapon(name, tier, move, getIcon(10, 7));
    }

    public static Weapon createStaffOfPower() {
        final String name = "Staff Of Power";
        final int tier = 3;
        final Move move = MoveFactory.createMagicMissileMove();

        return new Weapon(name, tier, move, getIcon(10, 6));
    }

    // boss weapons

    // Rock golem moves
    public static Weapon createRockHeal() {
        final String name = "Rock heal";
        final int tier = 1;
        final Move move = MoveFactory.createHealMove();

        return new Weapon(name, tier, move, getIcon(10, 6));
    }

    public static Weapon createRockSlam() {
        final String name = "Rock Slam";
        final int tier = 3;
        final Move move = MoveFactory.createSmashMove();

        return new Weapon(name, tier, move, getIcon(10, 6));
    }

    public static Weapon createRockThrow() {
        final String name = "Rock throw";
        final int tier = 3;
        final Move move = MoveFactory.createShootMove();

        return new Weapon(name, tier, move, getIcon(10, 6));
    }

    public static Weapon createRockCut() {
        final String name = "Rock Cut";
        final int tier = 3;
        final Move move = MoveFactory.createSlashMove();

        return new Weapon(name, tier, move, getIcon(10, 6));

    }

    // Ice golem moves

    public static Weapon creatFrostHeal() {
        final String name = "Frost heal";
        final int tier = 1;
        final Move move = MoveFactory.createHealMove();

        return new Weapon(name, tier, move, getIcon(10, 6));
    }

    public static Weapon createIceball() {
        final String name = "Ice ball";
        final int tier = 3;
        final Move move = MoveFactory.createMagicMissileMove();

        return new Weapon(name, tier, move, getIcon(10, 6));
    }

    public static Weapon createIceSpear() {
        final String name = "Ice spear attack";
        final int tier = 3;
        final Move move = MoveFactory.createSlashMove();

        return new Weapon(name, tier, move, getIcon(10, 6));
    }

    public static Weapon createIceClaws() {
        final String name = "Ice Claws";
        final int tier = 3;
        final Move move = MoveFactory.createSlashMove();

        return new Weapon(name, tier, move, getIcon(10, 6));

    }

    // Fire golem moves

    public static Weapon creatFireHeal() {
        final String name = "Fire heal";
        final int tier = 1;
        final Move move = MoveFactory.createHealMove();

        return new Weapon(name, tier, move, getIcon(10, 6));
    }

    public static Weapon createFireBall() {
        final String name = "Fireball";
        final int tier = 3;
        final Move move = MoveFactory.createMagicMissileMove();

        return new Weapon(name, tier, move, getIcon(10, 6));
    }

    public static Weapon createFireSword() {
        final String name = "Fire sword attack";
        final int tier = 3;
        final Move move = MoveFactory.createSlashMove();

        return new Weapon(name, tier, move, getIcon(10, 6));
    }

    public static Weapon createFireClaws() {
        final String name = "Fire Claws";
        final int tier = 3;
        final Move move = MoveFactory.createSlashMove();

        return new Weapon(name, tier, move, getIcon(10, 6));

    }
}

File: web/src/game/items/weapons/WeaponFactory.ts
import { weaponNamed, weaponSprite } from '../../utils/icons';
import { MoveFactory } from './moves/MoveFactory';
import { Weapon } from './Weapon';

export const WeaponFactory = {
  createRatClaws(): Weapon {
    return new Weapon('Rat Claws', 1, MoveFactory.createScratchMove(), weaponSprite(2, 4));
  },
  createMushroomCap(): Weapon {
    return new Weapon('Mushroom Cap', 1, MoveFactory.createSporeShotMove(), weaponNamed('Mushroom Cap'));
  },
  createBirdTalons(): Weapon {
    return new Weapon('Bird Talons', 1, MoveFactory.createScratchMove(), null);
  },
  createCrudeBow(): Weapon {
    return new Weapon('Crude Bow', 1, MoveFactory.createShootMove(), weaponSprite(9, 2));
  },
  createLongBow(): Weapon {
    return new Weapon('Long Bow', 2, MoveFactory.createShootMove(), weaponSprite(9, 3));
  },
  createCrossbow(): Weapon {
    return new Weapon('Crossbow', 3, MoveFactory.createShootMove(), weaponSprite(9, 4));
  },
  createToxicStaff(): Weapon {
    return new Weapon('Toxic Staff', 3, MoveFactory.createToxicBoltMove(), weaponSprite(10, 2));
  },
  createHealStaff(): Weapon {
    return new Weapon('Heal Staff', 3, MoveFactory.createHealMove(), weaponSprite(10, 1));
  },
  createRustyDagger(): Weapon {
    return new Weapon('Rusty Dagger', 1, MoveFactory.createSlashMove(), weaponSprite(0, 0));
  },
  createSteelSword(): Weapon {
    return new Weapon('Steel Sword', 2, MoveFactory.createSlashMove(), weaponSprite(0, 2));
  },
  createRoyalSword(): Weapon {
    return new Weapon('Royal Sword', 3, MoveFactory.createSlashMove(), weaponSprite(0, 8));
  },
  createWoodClub(): Weapon {
    return new Weapon('Wood Club', 1, MoveFactory.createSmashMove(), weaponSprite(8, 2));
  },
  createSteelMace(): Weapon {
    return new Weapon('Steel Mace', 2, MoveFactory.createSmashMove(), weaponSprite(5, 0));
  },
  createSteelHammer(): Weapon {
    return new Weapon('Steel Hammer', 3, MoveFactory.createSmashMove(), weaponSprite(4, 4));
  },
  createRudimentaryStaff(): Weapon {
    return new Weapon('Rudimentary Staff', 1, MoveFactory.createMagicMissileMove(), weaponSprite(10, 0));
  },
  createEnchantedStaff(): Weapon {
    return new Weapon('Enchanted Staff', 2, MoveFactory.createMagicMissileMove(), weaponSprite(10, 7));
  },
  createStaffOfPower(): Weapon {
    return new Weapon('Staff Of Power', 3, MoveFactory.createMagicMissileMove(), weaponSprite(10, 6));
  },
  createRockHeal(): Weapon {
    return new Weapon('Rock heal', 1, MoveFactory.createHealMove(), weaponSprite(10, 6));
  },
  createRockSlam(): Weapon {
    return new Weapon('Rock Slam', 3, MoveFactory.createSmashMove(), weaponSprite(10, 6));
  },
  createRockThrow(): Weapon {
    return new Weapon('Rock throw', 3, MoveFactory.createShootMove(), weaponSprite(10, 6));
  },
  createRockCut(): Weapon {
    return new Weapon('Rock Cut', 3, MoveFactory.createSlashMove(), weaponSprite(10, 6));
  },
  creatFrostHeal(): Weapon {
    return new Weapon('Frost heal', 1, MoveFactory.createHealMove(), weaponSprite(10, 6));
  },
  createIceball(): Weapon {
    return new Weapon('Ice ball', 3, MoveFactory.createMagicMissileMove(), weaponSprite(10, 6));
  },
  createIceSpear(): Weapon {
    return new Weapon('Ice spear attack', 3, MoveFactory.createSlashMove(), weaponSprite(10, 6));
  },
  createIceClaws(): Weapon {
    return new Weapon('Ice Claws', 3, MoveFactory.createSlashMove(), weaponSprite(10, 6));
  },
  creatFireHeal(): Weapon {
    return new Weapon('Fire heal', 1, MoveFactory.createHealMove(), weaponSprite(10, 6));
  },
  createFireBall(): Weapon {
    return new Weapon('Fireball', 3, MoveFactory.createMagicMissileMove(), weaponSprite(10, 6));
  },
  createFireSword(): Weapon {
    return new Weapon('Fire sword attack', 3, MoveFactory.createSlashMove(), weaponSprite(10, 6));
  },
  createFireClaws(): Weapon {
    return new Weapon('Fire Claws', 3, MoveFactory.createSlashMove(), weaponSprite(10, 6));
  },
};
